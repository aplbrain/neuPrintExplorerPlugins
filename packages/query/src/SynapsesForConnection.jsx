import React from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router';

import randomColor from 'randomcolor';

import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import FormControl from '@material-ui/core/FormControl';
import { withStyles } from '@material-ui/core/styles';
import { round } from 'mathjs';
import Select from 'react-select';
import { setColumnIndices } from './shared/pluginhelpers';

const pluginName = 'SynapsesForConnection';

const styles = theme => ({
  textField: {
    width: 300,
    margin: '0 0 1em 0'
  },
  button: {
    margin: 4,
    display: 'block'
  },
  formControl: {},
  select: {
    fontFamily: theme.typography.fontFamily,
    margin: '0.5em 0 1em 0'
  }
});

const pluginAbbrev = 'sfc';

export class SynapsesForConnection extends React.Component {
  static get queryName() {
    return 'Synapses for connection';
  }

  static get queryDescription() {
    return 'Retrieves synapses involved in a connection.';
  }

  static get isExperimental() {
    return true;
  }

  static get queryAbbreviation() {
    return pluginAbbrev;
  }

  processRequest = () => {
    const { dataSet, actions, history } = this.props;
    const { bodyId1 = '', bodyId2 = '', rois = [] } = actions.getQueryObject(pluginAbbrev);

    let roiPredicate = '';
    if (rois.length > 0) {
      roiPredicate = ' WHERE (';
      rois.forEach(roi => {
        roiPredicate += `exists(s.\`${roi}\`) AND `;
      });
      roiPredicate = roiPredicate.slice(0, -5);
      roiPredicate += ')';
    }

    const cypherQuery = `MATCH (a:\`${dataSet}-Neuron\`{bodyId:${bodyId1}})<-[:From]-(c:ConnectionSet)-[:To]->(b{bodyId:${bodyId2}}), (c)-[:Contains]->(s:Synapse)${roiPredicate} RETURN s.type, s.location.x ,s.location.y ,s.location.z, s.confidence, keys(s)`;

    const query = {
      dataSet,
      cypherQuery,
      visType: 'SimpleTable',
      plugin: pluginName,
      parameters: { bodyId1, bodyId2, rois },
      title: `Synapses involved in connection between ${bodyId1} and ${bodyId2}`,
      menuColor: randomColor({ luminosity: 'light', hue: 'random' }),
      processResults: this.processResults
    };
    actions.submit(query);
    history.push({
      pathname: '/results',
      search: actions.getQueryString()
    });
    return query;
  };

  processResults = (query, apiResponse) => {
    const { actions } = this.props;
    const { parameters = {} } = query;
    const { bodyId1 = '', bodyId2 = '', rois = [] } = parameters;
    const indexOf = setColumnIndices(['type', 'location', 'confidence', 'rois']);

    if (apiResponse.data.length > 0) {
      const data = apiResponse.data.map(row => {
        const type = row[0];
        const x = row[1];
        const y = row[2];
        const z = row[3];
        const confidence = row[4];
        const properties = row[5];

        const converted = [];
        converted[indexOf.type] = type;
        converted[indexOf.location] = `[${x},${y},${z}]`;
        converted[indexOf.confidence] = round(confidence, 4);
        converted[indexOf.rois] = properties.filter(
          value =>
            value !== 'type' &&
            value !== 'location' &&
            value !== 'confidence' &&
            value !== 'timeStamp'
        );

        return converted;
      });

      const columns = [];
      columns[indexOf.type] = 'type';
      columns[indexOf.location] = 'location';
      columns[indexOf.confidence] = 'confidence';
      columns[indexOf.rois] = 'rois';

      return {
        columns,
        data,
        debug: apiResponse.debug
      };
    }

    actions.pluginResponseError(
      `No synapses between ${bodyId1} and ${bodyId2} in specified rois (${rois})`
    );
    return {
      columns: [],
      data: [],
      debug: apiResponse.debug
    };
  };

  addBodyId1 = event => {
    const { actions } = this.props;
    actions.setQueryString({
      [pluginAbbrev]: {
        bodyId1: event.target.value
      }
    });
  };

  addBodyId2 = event => {
    const { actions } = this.props;
    actions.setQueryString({
      [pluginAbbrev]: {
        bodyId2: event.target.value
      }
    });
  };

  handleChangeRois = selected => {
    const { actions } = this.props;
    const rois = selected.map(item => item.value);
    actions.setQueryString({
      [pluginAbbrev]: {
        rois
      }
    });
  };

  render() {
    const { actions, classes, isQuerying, availableROIs } = this.props;
    const { bodyId1 = '', bodyId2 = '', rois = [] } = actions.getQueryObject(pluginAbbrev);

    const roiOptions = availableROIs.map(name => ({
      label: name,
      value: name
    }));

    const roiValues = rois.map(roi => ({
      label: roi,
      value: roi
    }));

    return (
      <FormControl className={classes.formControl}>
        <TextField
          label="Neuron ID A"
          multiline
          fullWidth
          rows={1}
          value={bodyId1}
          rowsMax={1}
          className={classes.textField}
          onChange={this.addBodyId1}
        />
        <TextField
          label="Neuron ID B"
          multiline
          fullWidth
          rows={1}
          value={bodyId2}
          rowsMax={1}
          className={classes.textField}
          onChange={this.addBodyId2}
        />
        <Select
          className={classes.select}
          isMulti
          value={roiValues}
          onChange={this.handleChangeRois}
          options={roiOptions}
          closeMenuOnSelect={false}
        />
        <Button
          variant="contained"
          className={classes.button}
          onClick={this.processRequest}
          color="primary"
          disabled={isQuerying}
        >
          Submit
        </Button>
      </FormControl>
    );
  }
}

SynapsesForConnection.propTypes = {
  classes: PropTypes.object.isRequired,
  actions: PropTypes.object.isRequired,
  dataSet: PropTypes.string.isRequired,
  history: PropTypes.object.isRequired,
  isQuerying: PropTypes.bool.isRequired,
  availableROIs: PropTypes.arrayOf(PropTypes.string).isRequired
};

export default withStyles(styles)(withRouter(SynapsesForConnection));
