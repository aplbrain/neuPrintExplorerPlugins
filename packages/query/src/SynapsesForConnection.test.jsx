import { SynapsesForConnection } from './SynapsesForConnection';

let wrapper;
let button;
let bodyAField;
let bodyBField;
let roiSelect;

const emptyApiResponse = {
  columns: ['s.type', 's.location.x', 's.location.y', 's.location.z', 's.confidence', 'keys(s)'],
  data: [],
  debug: 'test'
};
const query = { parameters: { bodyId1: '123456', bodyId2: '645321', rois: [] } };
const apiResponse = {
  columns: ['s.type', 's.location.x', 's.location.y', 's.location.z', 's.confidence', 'keys(s)'],
  data: [
    [
      'pre',
      1.1,
      2.1,
      3.1,
      0.9839201,
      ['type', 'location', 'timeStamp', 'confidence', 'roi1', 'roi2']
    ],
    ['post', 1.0, 3.0, 3.0, 0.1, ['type', 'location', 'timeStamp', 'confidence']]
  ],
  debug: 'test'
};

const styles = { select: {}, clickable: {} };
const { actions, React, enzyme, renderer, submit } = global;

const neoServerSettings = {
  get: () => 'http://example.com'
};

const component = (
  <SynapsesForConnection
    availableROIs={['roiA', 'roiB', 'roiC']}
    dataSet="test"
    datasetstr="test"
    actions={actions}
    submit={submit}
    classes={styles}
    history={{ push: jest.fn() }}
    isQuerying={false}
    neoServerSettings={neoServerSettings}
    neoServer="testServer"
  />
);

describe('synapses for connection Plugin', () => {
  beforeAll(() => {
    wrapper = enzyme.mount(component);
    button = wrapper.find('Button');
    bodyAField = wrapper.find('TextField').at(0);
    bodyBField = wrapper.find('TextField').at(1);
    roiSelect = wrapper.find('Select');
  });
  beforeEach(() => {
    submit.mockClear();
  });
  it('has name and description', () => {
    expect(SynapsesForConnection.details.name).toBeTruthy();
    expect(SynapsesForConnection.details.description).toBeTruthy();
  });
  it('renders correctly', () => {
    const pluginView = renderer.create(component).toJSON();
    expect(pluginView).toMatchSnapshot();
  });
  describe('when user clicks submit', () => {
    it('should return a query object with input fields contained in cypher string and should submit', () => {
      bodyAField.props().onChange({ target: { value: '123456' } });
      bodyBField.props().onChange({ target: { value: '645321' } });

      expect(button.props().onClick()).toEqual(undefined);
      expect(submit).toHaveBeenCalledTimes(1);

      roiSelect.props().onChange([{ value: 'roi1' }, { value: 'roi2' }]);

      expect(button.props().onClick()).toEqual(undefined);
      expect(submit).toHaveBeenCalledTimes(2);
    });
  });

  describe('processes returned results', () => {
    it('should produce error, empty data object if results are empty', () => {
      const processedEmptyResults = SynapsesForConnection.processResults(
        {},
        emptyApiResponse,
        actions
      );
      expect(actions.pluginResponseError).toHaveBeenCalledTimes(1);
      expect(processedEmptyResults).toEqual({
        columns: [],
        data: [],
        debug: emptyApiResponse.debug,
        title: 'Synapses involved in connection between  and '
      });
    });
    it('should produce object with data rows', () => {
      const processedResults = SynapsesForConnection.processResults(query, apiResponse, actions);
      const { columns, data, debug } = processedResults;
      expect(columns).toEqual(['type', 'location', 'confidence', 'rois']);
      expect(data[0]).toEqual(['pre', '[1.1,2.1,3.1]', 0.9839, ['roi1', 'roi2']]);
      expect(data.length).toBe(2);
      expect(debug).toBe(apiResponse.debug);
    });
  });
});
