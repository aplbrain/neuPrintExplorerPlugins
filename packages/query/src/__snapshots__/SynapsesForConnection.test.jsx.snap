// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`synapses for connection Plugin renders correctly 1`] = `
<div
  className="MuiFormControl-root-1"
>
  <div
    className="MuiFormControl-root-1 MuiFormControl-fullWidth-4"
  >
    <label
      className="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13"
      data-shrink={false}
    >
      Neuron ID A
    </label>
    <div
      className="MuiInputBase-root-36 MuiInput-root-23 MuiInput-underline-27 MuiInputBase-fullWidth-45 MuiInput-fullWidth-30 MuiInputBase-formControl-37 MuiInput-formControl-24 MuiInputBase-multiline-44 MuiInput-multiline-29"
      onClick={[Function]}
    >
      <div
        className="MuiPrivateTextarea-root-53"
      >
        <textarea
          aria-hidden="true"
          className="MuiPrivateTextarea-textarea-54 MuiPrivateTextarea-shadow-55"
          readOnly={true}
          rows="1"
          tabIndex={-1}
          value=""
        />
        <textarea
          aria-hidden="true"
          className="MuiPrivateTextarea-textarea-54 MuiPrivateTextarea-shadow-55"
          readOnly={true}
          rows={1}
          tabIndex={-1}
          value=""
        />
        <textarea
          aria-invalid={false}
          className="MuiPrivateTextarea-textarea-54 MuiInputBase-input-46 MuiInput-input-31 MuiInputBase-inputMultiline-48 MuiInput-inputMultiline-33"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          rows={1}
          style={
            Object {
              "height": 19,
            }
          }
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="MuiFormControl-root-1 MuiFormControl-fullWidth-4"
  >
    <label
      className="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13"
      data-shrink={false}
    >
      Neuron ID B
    </label>
    <div
      className="MuiInputBase-root-36 MuiInput-root-23 MuiInput-underline-27 MuiInputBase-fullWidth-45 MuiInput-fullWidth-30 MuiInputBase-formControl-37 MuiInput-formControl-24 MuiInputBase-multiline-44 MuiInput-multiline-29"
      onClick={[Function]}
    >
      <div
        className="MuiPrivateTextarea-root-53"
      >
        <textarea
          aria-hidden="true"
          className="MuiPrivateTextarea-textarea-54 MuiPrivateTextarea-shadow-55"
          readOnly={true}
          rows="1"
          tabIndex={-1}
          value=""
        />
        <textarea
          aria-hidden="true"
          className="MuiPrivateTextarea-textarea-54 MuiPrivateTextarea-shadow-55"
          readOnly={true}
          rows={1}
          tabIndex={-1}
          value=""
        />
        <textarea
          aria-invalid={false}
          className="MuiPrivateTextarea-textarea-54 MuiInputBase-input-46 MuiInput-input-31 MuiInputBase-inputMultiline-48 MuiInput-inputMultiline-33"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          rows={1}
          style={
            Object {
              "height": 19,
            }
          }
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="css-10nd86i [object Object]"
    onKeyDown={[Function]}
  >
    <div
      className="css-vj8t7z"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className="css-1hwfws3"
      >
        <div
          className="css-1492t68"
        >
          Select...
        </div>
        <div
          className="css-1g6gooi"
        >
          <div
            className=""
            style={
              Object {
                "display": "inline-block",
              }
            }
          >
            <input
              aria-autocomplete="list"
              autoCapitalize="none"
              autoComplete="off"
              autoCorrect="off"
              disabled={false}
              id="react-select-3-input"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              spellCheck="false"
              style={
                Object {
                  "background": 0,
                  "border": 0,
                  "boxSizing": "content-box",
                  "color": "inherit",
                  "fontSize": "inherit",
                  "opacity": 1,
                  "outline": 0,
                  "padding": 0,
                  "width": "1px",
                }
              }
              tabIndex="0"
              type="text"
              value=""
            />
            <div
              style={
                Object {
                  "height": 0,
                  "left": 0,
                  "overflow": "scroll",
                  "position": "absolute",
                  "top": 0,
                  "visibility": "hidden",
                  "whiteSpace": "pre",
                }
              }
            >
              
            </div>
          </div>
        </div>
      </div>
      <div
        className="css-1wy0on6"
      >
        <span
          className="css-d8oujb"
        />
        <div
          aria-hidden="true"
          className="css-1ep9fjw"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <svg
            aria-hidden="true"
            className="css-19bqh2r"
            focusable="false"
            height={20}
            viewBox="0 0 20 20"
            width={20}
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <button
    className="MuiButtonBase-root-82 MuiButton-root-56 MuiButton-contained-67 MuiButton-containedPrimary-68 MuiButton-raised-70 MuiButton-raisedPrimary-71"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiButton-label-57"
    >
      Submit
    </span>
    <span
      className="MuiTouchRipple-root-85"
    />
  </button>
</div>
`;
